File: src/lib/git/pull.ml
open Common

let remotes _ = 1

let run ~mgr ~path ~(ctx : Miroir.Context.context) ~disp ~slot ~sem:_ ~args =
  let name = repo_name path in
  Miroir.Display.repo disp slot (Printf.sprintf "%s :: pull" name);
  let* () = ensure_repo path in
  (* pull uses origin remote, index 0 *)
  let j = 0 in
  let out = Miroir.Display.output disp slot j in
  let info s = Miroir.Display.remote disp slot j s in
  info "pulling...";
  let* () =
    run
      ~mgr
      ~cwd:path
      ~env:ctx.env
      ~on_output:out
      ([ "pull"; "origin"; ctx.branch ] @ args)
  in
  info "updating submodules...";
  let r =
    run
      ~mgr
      ~cwd:path
      ~env:ctx.env
      ~on_output:out
      [ "submodule"; "update"; "--recursive"; "--remote" ]
  in
  (match r with
   | Ok () -> info "done"
   | Error e -> info (Printf.sprintf "error: %s" e));
  r
;;

File: internal/git/pull.go
package git

import (
	"errors"
	"fmt"
)

type Pull struct{}

func (Pull) Remotes(_ int) int { return 1 }

func (Pull) Run(p Params) error {
	name := repoName(p.Path)
	p.Disp.Repo(p.Slot, fmt.Sprintf("%s :: pull", name))
	if err := ensureRepo(p.Path); err != nil {
		return err
	}

	j := 0
	out := func(s string) { p.Disp.Output(p.Slot, j, s) }
	info := func(s string) { p.Disp.Remote(p.Slot, j, s) }

	if !p.Force && isDirty(p.Path, p.Ctx.Env) {
		msg := "dirty working tree, use --force to override"
		p.Disp.ErrorRemote(p.Slot, j, fmt.Sprintf("error: %s", msg))
		return errors.New(msg)
	}

	if p.Force {
		info("resetting...")
		if err := run(p.Path, p.Ctx.Env, true, nil,
			"reset", "--hard", "HEAD"); err != nil {
			return err
		}
	}

	info("pulling...")
	pullArgs := append([]string{"pull", "origin", p.Ctx.Branch}, p.Args...)
	if err := run(p.Path, p.Ctx.Env, false, out, pullArgs...); err != nil {
		return err
	}

	info("updating submodules...")
	err := run(p.Path, p.Ctx.Env, false, out,
		"submodule", "update", "--recursive", "--init")
	if err != nil {
		p.Disp.ErrorRemote(p.Slot, j, fmt.Sprintf("error: %s", err))
	} else {
		info("done")
	}
	return err
}
